File: apps/desktop/src/lib/localClassification.ts
import { invoke } from "@tauri-apps/api/core";
import type { AppSettings } from "./types";

export interface LocalClassificationStatus {
  supported: boolean;
  installed: boolean;
  enabled: boolean;
  busy: boolean;
  phase: string;
  done: number;
  total: number;
  failed: number;
  download_done: number;
  download_total: number;
  message: string;
  last_error: string;
  model: string;
  acceleration: string;
  vector_supported: boolean;
  vector_installed: boolean;
}
export interface ClassificationLabel {
  id: string;
  name: string;
  description: string;
  enabled: boolean;
  count: number;
  has_examples: boolean;
}
export const localClassification = {
  status: () => invoke<LocalClassificationStatus>("local_classification_status"),
  labels: () => invoke<ClassificationLabel[]>("local_classification_labels"),
  start: (options: { install?: boolean; tagId?: string; pendingOnly?: boolean } = {}) =>
    invoke<void>("local_classification_start", { install: false, tagId: null, pendingOnly: true, ...options }),
  stop: () => invoke<void>("local_classification_stop"),
  vectorInstall: () => invoke<void>("local_classification_vector_install"),
  enable: (enabled: boolean) => invoke<void>("local_classification_enable", { enabled }),
  save: (id: string | null, name: string, description: string, enabled: boolean) =>
    invoke<string>("local_classification_save_label", { id, name, description, enabled }),
  example: (tagId: string, assetIds: string[], positive: boolean) =>
    invoke<void>("local_classification_example", { tagId, assetIds, positive }),
  jevConfig: async (): Promise<{ enabled: boolean; apiKey: string }> => {
    const settings = await invoke<AppSettings>("get_settings");
    return { enabled: settings.jev_verify_enabled, apiKey: settings.jev_api_key ?? "" };
  },
  saveJevConfig: async (enabled: boolean, apiKey: string) => {
    const settings = await invoke<AppSettings>("get_settings");
    settings.jev_verify_enabled = enabled;
    settings.jev_api_key = apiKey.trim() ? apiKey.trim() : null;
    await invoke<void>("update_settings", { settings });
  },
};

File: apps/desktop/src-tauri/src/commands/local_classification.rs
use crate::core::local_classification::{data::Label, LocalClassifier, Status};
use crate::core::paths::LibraryPaths;
use crate::core::settings::SettingsState;
use crate::db::Database;
use std::sync::Arc;
use tauri::{AppHandle, Emitter, State};

#[tauri::command]
pub async fn local_classification_status(
    classifier: State<'_, Arc<LocalClassifier>>,
    db: State<'_, Arc<Database>>,
) -> Result<Status, String> {
    classifier.snapshot(&db)
}

#[tauri::command]
pub async fn local_classification_start(
    app: AppHandle,
    classifier: State<'_, Arc<LocalClassifier>>,
    db: State<'_, Arc<Database>>,
    paths: State<'_, Arc<LibraryPaths>>,
    settings: State<'_, SettingsState>,
    install: bool,
    tag_id: Option<String>,
    pending_only: bool,
) -> Result<(), String> {
    let snapshot = settings.get();
    let jev_key = if snapshot.jev_verify_enabled {
        snapshot.jev_api_key.filter(|key| !key.trim().is_empty())
    } else {
        None
    };
    classifier.inner().start(
        app,
        db.inner().clone(),
        paths.inner().clone(),
        install,
        tag_id,
        jev_key,
        pending_only,
    )
}

#[tauri::command]
pub async fn local_classification_vector_install(
    app: AppHandle,
    classifier: State<'_, Arc<LocalClassifier>>,
) -> Result<(), String> {
    classifier.inner().start_vector(app)
}

#[tauri::command]
pub async fn local_classification_vector_uninstall(
    classifier: State<'_, Arc<LocalClassifier>>,
) -> Result<(), String> {
    classifier.inner().uninstall_vector()
}

#[tauri::command]
pub async fn local_classification_stop(
    classifier: State<'_, Arc<LocalClassifier>>,
    db: State<'_, Arc<Database>>,
) -> Result<(), String> {
    db.set_local_enabled(false).map_err(|e| e.to_string())?;
    classifier.cancel();
    Ok(())
}

#[tauri::command]
pub async fn local_classification_enable(
    db: State<'_, Arc<Database>>,
    enabled: bool,
) -> Result<(), String> {
    db.set_local_enabled(enabled).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn local_classification_labels(
    db: State<'_, Arc<Database>>,
) -> Result<Vec<Label>, String> {
    db.local_labels().map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn local_classification_save_label(
    app: AppHandle,
    db: State<'_, Arc<Database>>,
    id: Option<String>,
    name: String,
    description: String,
    enabled: bool,
) -> Result<String, String> {
    let result = db
        .save_local_label(id.as_deref(), &name, &description, enabled)
        .map_err(|e| e.to_string())?;
    let _ = app.emit("library://assets-changed", ());
    Ok(result)
}

#[tauri::command]
pub async fn local_classification_example(
    app: AppHandle,
    db: State<'_, Arc<Database>>,
    tag_id: String,
    asset_ids: Vec<String>,
    positive: bool,
) -> Result<(), String> {
    db.local_example(&tag_id, &asset_ids, positive)
        .map_err(|e| e.to_string())?;
    let _ = app.emit("library://assets-changed", ());
    Ok(())
}
